### PR TITLE
Forum application complete version

### PR DESCRIPTION
**Pull request by jmalicevic**
This is the most complete version of the app so far. Note the following:
* Vote extensions are currently only processed by the proposer
* FinalizeBlock does not update the DB (it reads from it though); but puts transaction commands within one transaction
* Commit commits the transaction 
* Curse words are located in the `app.toml` file

### CODE REVIEW

File: abci/app.go [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+func (app ForumApp) Query(ctx context.Context, query *abci.RequestQuery) (*abci.ResponseQuery, error) {
+	resp := abci.ResponseQuery{Key: query.Data}
+	// Just testing whether validators are properly stored
+	if query.Path == "/val" {
+		for k := range app.valAddrToPubKeyMap {
+			return &types.ResponseQuery{
+				Key:   query.Data,
+				Value: []byte(string(k)),
+			}, nil
+		}
```
Comment (sergio-mena) [inline abci/app.go]: Won't you return the `ResponseQuery` at the first iteration of the loop, no matter how many validators there are?

Comment (sergio-mena) [reply abci/app.go]: I guess there's an "if" missing...

File: abci/app.go [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+
+	return &resp, nil
+}
+
+func (app ForumApp) CheckTx(ctx context.Context, checktx *abci.RequestCheckTx) (*abci.ResponseCheckTx, error) {
+	// Parse the tx message
+	msg, err := model.ParseMessage(checktx.Tx)
+	if err != nil {
+		fmt.Printf("failed to parse transaction message checktx: %v\n", err)
+		return &abci.ResponseCheckTx{Code: CodeTypeInvalidTxFormat, Log: "Invalid transaction format"}, err
```
Comment (sergio-mena) [inline abci/app.go]:
```suggestion
		return &abci.ResponseCheckTx{Code: CodeTypeInvalidTxFormat, Log: "Invalid transaction format"}, nil
```
You shouldn't return errors in ABCI calls. You're already returning the error code in `ResponseCheckTx`.
If you want the client to see the contents of `err`, you can append it to the `Log` message.

File: abci/app.go [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+		fmt.Println("problem in check tx: ", string(checktx.Tx))
+		return &types.ResponseCheckTx{Code: CodeTypeEncodingError}, nil
+	}
+
+	if err == badger.ErrKeyNotFound {
+		fmt.Println("Not found user :", msg.Sender)
+	}
+	if u != nil && u.Banned {
+		return &types.ResponseCheckTx{Code: CodeTypeBanned, Log: "User is banned"}, nil
+	}
```
Comment (sergio-mena) [inline abci/app.go]:
```suggestion
	if err != nil {
		if !errors.Is(err, badger.ErrKeyNotFound) {
			fmt.Println("problem in check tx: ", string(checktx.Tx))
			return &types.ResponseCheckTx{Code: CodeTypeEncodingError}, nil
		}
		fmt.Println("Not found user :", msg.Sender)
	} else {
		if u != nil && u.Banned {
			return &types.ResponseCheckTx{Code: CodeTypeBanned, Log: "User is banned"}, nil
		}
	}
```
I find these conditions hard to follow, and hard to convince me they are exhaustive. This is a suggestion to make conditions clearer

File: abci/app.go [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+}
+
+// Consensus Connection
+// Initialize blockchain w validators/other info from CometBFT
+func (app ForumApp) InitChain(_ context.Context, req *abci.RequestInitChain) (*abci.ResponseInitChain, error) {
+	for _, v := range req.Validators {
+		app.updateValidator(v)
+	}
+	appHash := make([]byte, 8)
+	binary.PutVarint(appHash, app.state.Size)
```
Comment (sergio-mena) [inline abci/app.go]: ????

File: abci/app.go [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+	req.ConsensusParams.Abci.VoteExtensionsEnableHeight = 1
```
Comment (sergio-mena) [inline abci/app.go:146]: You might want to say you can also do this in directly in the genesis file (that's e.g. what the e2e tests do)

File: abci/app.go [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+
+	// prepare proposal puts the BanTx first, then adds the other transactions
+	// ProcessProposal should verify this
+	proposedTxs := make([][]byte, 0)
+	finalProposal := make([][]byte, 0)
+	bannedUsersString := make(map[string]struct{})
+	for _, tx := range proposal.Txs {
+		msg, err := model.ParseMessage(tx)
+		if err == nil {
+			if err != nil {
```
Comment (sergio-mena) [inline abci/app.go]: This can't ever be true, since you're checking the opposite in line 167

File: abci/app.go [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+			// Adding the curse words from vote extensions too
+			if !IsCurseWord(msg.Message, voteExtensionCurseWords) {
+				proposedTxs = append(proposedTxs, tx)
+			} else {
+				banTx := model.BanTx{UserName: msg.Sender}
+				bannedUsersString[msg.Message] = struct{}{}
+				resultBytes, err := json.Marshal(banTx)
+				if err == nil {
+					finalProposal = append(finalProposal, resultBytes)
+				}
```
Comment (sergio-mena) [inline abci/app.go:180]: So what do we do if this error is not `nil`?

Comment (sergio-mena) [inline abci/app.go]: So what do we do if this error is not `nil`?

File: abci/app.go [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+			}
+		}
+	}
+	// Need to loop again through the proposed Txs to make sure there is none left by a user that was banned after the tx was accepted
+	for _, tx := range proposedTxs {
+		// there should be no error here as these are just transactions we have checked and added
+		msg, _ := model.ParseMessage(tx)
```
Comment (sergio-mena) [inline abci/app.go]: If we're sure there won't be any errors, we should add an "if" checking it and panic if there happens to be an error.
This prevents a possible bug from further propagating. Also, the panic would provide a handy stack trace

File: abci/app.go [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+		if _, ok := bannedUsersString[msg.Sender]; !ok {
```
Comment (sergio-mena) [inline abci/app.go:187]: Just double checking... you're looking up `msg.Sender` here, but you set `msg.Message` above (line 176). I'm not sure I get it

File: abci/app.go [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+		var parsedBan model.BanTx
+		var parsedTx model.Message
+		var err error
+		if !processedBanTxs && isBanTx(tx) {
+			fmt.Println("FoundBanTx")
+			err = json.Unmarshal(tx, &parsedBan)
+			if err != nil {
+				return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_REJECT}, err
+			}
+			if processedBanTxs {
```
Comment (sergio-mena) [inline abci/app.go]: This can never be true because of line 203

File: abci/app.go [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+				// Banning transactions have to come first, cannot have them once we hit the first non user ban tx
+				return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_REJECT}, err
+			}
+			bannedUsers[parsedBan.UserName] = struct{}{}
+		} else {
```
Comment (sergio-mena) [inline abci/app.go:207]: Not clear what type of TX we're processing here, "ban" or "regular"?

File: abci/app.go [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+			_, err = model.ParseMessage(tx)
+			if err != nil {
+				return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_REJECT}, err
+			}
+			processedBanTxs = true
```
Comment (sergio-mena) [inline abci/app.go]: So, we're only able to process one ban tx?

File: abci/app.go [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+}
+
+func (ForumApp) ApplySnapshotChunk(_ context.Context, applysnapshotchunk *abci.RequestApplySnapshotChunk) (*abci.ResponseApplySnapshotChunk, error) {
+	return &abci.ResponseApplySnapshotChunk{}, nil
+}
+
+func (app ForumApp) ExtendVote(_ context.Context, extendvote *abci.RequestExtendVote) (*abci.ResponseExtendVote, error) {
+	fmt.Println("Entered extend vote")
+
+	return &abci.ResponseExtendVote{VoteExtension: []byte(app.CurseWords)}, nil
```
Comment (sergio-mena) [inline abci/app.go:327]: We should check somewhere (InitChain?) that we're not fed a config file with repeated curse words

File: model/db.go [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+		}
+		err = txn.Set([]byte(user.Name), userBytes)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
```
Comment (sergio-mena) [inline model/db.go:78]:
```suggestion
```

File: model/db.go [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+	return nil
```
Comment (sergio-mena) [inline model/db.go]:
```suggestion
	return err
```

File: model/db.go [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+		if err != nil {
+			return err
+		}
+		err = item.Value(func(val []byte) error {
+			return json.Unmarshal(val, &user)
+		})
+		return err
+	})
+	if err != nil {
+		fmt.Println("Error in retrieveing user: ", err)
```
Comment (sergio-mena) [inline model/db.go]:
```suggestion
		fmt.Println("Error in retrieving user: ", err)
```

File: model/db.go [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+			if err != badger.ErrKeyNotFound {
+				return err
+			}
+			return nil
+		}
+		value, err = item.ValueCopy(nil)
+		if err != nil {
+			return err
+		}
+		return nil
```
Comment (sergio-mena) [inline model/db.go]:
```suggestion
		return err
```

File: model/db.go [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+		defer it.Close()
+		for it.Rewind(); it.Valid(); it.Next() {
+			var err error
+			item := it.Item()
+			k := item.Key()
+			if isValidatorTx(k) {
+				err := item.Value(func(v []byte) error {
+					validator := new(types.ValidatorUpdate)
+					err = types.ReadMessage(bytes.NewBuffer(v), validator)
+					validators = append(validators, *validator)
```
Comment (sergio-mena) [inline model/db.go]: If there is an error in the line above, shouldn't we skip this one?

File: model/db.go [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+				if err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	})
+	return
+}
+func isValidatorTx(tx []byte) bool {
```
Comment (sergio-mena) [inline model/db.go:162]: I see we support validator TXs. Shouldn't `FinalizeBlock` be processing them?
(IMO, we should keep this simple, set the validators in genesis, set them in the state at `InitChain` and leave them untouched)

File: model/messages.go [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+// Message represents a message sent by a user
+type Message struct {
+	Sender  string `json:"sender"`
+	Message string `json:"message"`
+}
+
+type MsgHistory struct {
+	Msg string `json:"history"`
```
Comment (sergio-mena) [inline model/messages.go:22]: Is this the history of all messages? If so, we could have a first version of an appHash by hashing this message (I believe there is only one in the DB)

File: model/messages.go [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+func AppendToChat(db *DB, message Message) (string, error) {
+	historyBytes, err := ViewDB(db.GetDB(), []byte("history"))
+	if err != nil {
+		fmt.Println("Error fething historu:", err)
```
Comment (sergio-mena) [inline model/messages.go]:
```suggestion
		fmt.Println("Error fething history:", err)
```

File: model/messages.go [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+		return "", err
+	}
+	msgBytes := string(historyBytes)
```
Comment (sergio-mena) [inline model/messages.go]: You could replace all this code by just calling `FetchHistory`

File: model/user.go [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+	Moderator     bool
+	Banned        bool
+	NumMessages   int64
+	Version       uint64
+	SchemaVersion int
+}
+
+func (u *User) IsBanned() bool {
+	return u.Banned
+}
```
Comment (sergio-mena) [inline model/user.go]: Do we need this?

File: abci/app.go (hunk 1 above)
Comment (jmalicevic) [reply abci/app.go]: Yes, you are right, this was a quick test which I forgot to update. This should be actually read from the DB.

File: abci/app.go (hunk 2 above)
Comment (jmalicevic) [reply abci/app.go]: Indeed, missed to fix this one.

File: abci/app.go (hunk 4 above)
Comment (jmalicevic) [reply abci/app.go]: I copy pasted this from the [kvstore example](https://github.com/cometbft/cometbft/blob/028fb5eb0c071ec2c5fea34fe3d02b1fea814d6d/abci/example/kvstore/kvstore.go#LL117C1-L117C1) on main, I did not really give it much thought.  My guess is the point of it is to send the initial `appHash`. But  it could be done by calling actually `app.state.Hash()` as is done below. Will fix it.

File: abci/app.go (hunk 7 above)
Comment (jmalicevic) [reply abci/app.go]: We should probably panic because if we can't create this TX then we will not ban who we should ban. ToDo

File: abci/app.go (hunk 8 above)
Comment (jmalicevic) [reply abci/app.go]: I agree. Will add this.

File: abci/app.go (hunk 9 above)
Comment (jmalicevic) [reply abci/app.go:187]: This is the extra check I explained in the call. This loop is done after all transactions have been processed. And what can happen is :
`Message: {sender:bob; message:hi}` -> OK, add to `proposedTx`
`Message: {sender:bob; message:curse_workd}` -> notOK , create `BanTx{username:Bob}` ; add this to `finalProposal` and add `bob` to `bannedUserString`.


Now, if we just appended `proposedTx` to `finalProposal` bobs first message would get in. But we want to ban bob and to not want his messages. So we make sure that this block will not contain any message from bob. And `bannedUserString` is used to verify that the sender of a message that is already in `proposedTx` does not exist in this map. If it does not exist, means the user was not banned here.

File: abci/app.go (hunk 10 above)
Comment (jmalicevic) [reply abci/app.go]: Yes sorry, this was a leftover from previous versions. Thanks for the catch!

File: abci/app.go (hunk 11 above)
Comment (jmalicevic) [reply abci/app.go:207]: Regular. Meaning that, the transaction is a Ban transaction and we are not done processing them.

File: abci/app.go (hunk 12 above)
Comment (jmalicevic) [reply abci/app.go]: No, we will process as many ban transactions as they are. But they have to come one after the other - so before we hit this `else`. If they come after the `else` has been hit that means that the proposer did not put the ban transactions in front of the regular ones. (This is why in `PrepareProposal` banned transactions are appended directly to `finalProposal` and the regular ones are appended afterwards.

File: abci/app.go (hunk 13 above)
Comment (jmalicevic) [reply abci/app.go:327]: Yes this is one check I did not get a chance to include.

File: model/db.go (hunk 19 above)
Comment (jmalicevic) [reply model/db.go:162]: I agree. I think we can delete the val transactions actually. I added them because I wanted to confirm we save them correctly (to verify that the extensions come from validators we know about), but they are not essential for the demo. ?

File: model/messages.go (hunk 20 above)
Comment (jmalicevic) [reply model/messages.go:22]: Yes, that is the more correct way to do it. I followed what was done in the kvstore , but this would be more correct.

File: model/user.go (hunk 23 above)
Comment (jmalicevic) [reply model/user.go]: Nope

File: abci/app.go (hunk 9 above)
Comment (sergio-mena) [reply abci/app.go:187]: Talked online, `msg.Sender` `msg.Message` are inconsistent

File: model/db.go (hunk 19 above)
Comment (jmalicevic) [reply model/db.go:162]: Ok so I left the function here because when we add them to the sate we need to read them in. But I removed the option to update this via the RPC

File: abci/app.go [hunk 24]
```diff
@@ -0,0 +1,377 @@
...
+
+	// prepare proposal puts the BanTx first, then adds the other transactions
+	// ProcessProposal should verify this
+	proposedTxs := make([][]byte, 0)
+	finalProposal := make([][]byte, 0)
+	bannedUsersString := make(map[string]struct{})
+	for _, tx := range proposal.Txs {
+		msg, err := model.ParseMessage(tx)
+		if err == nil {
+			proposedTxs = append(proposedTxs, tx)
```
Comment (sergio-mena) [inline abci/app.go]: I don't get this, what if it is a curse word?
Shouldn't this rather be something like:
```
if err != nil {
    continue
}

File: abci/app.go [hunk 25]
```diff
@@ -0,0 +1,377 @@
...
+		if !IsCurseWord(msg.Message, voteExtensionCurseWords) {
+			proposedTxs = append(proposedTxs, tx)
+		} else {
+			banTx := model.BanTx{UserName: msg.Sender}
+			bannedUsersString[msg.Sender] = struct{}{}
+			resultBytes, err := json.Marshal(banTx)
+			if err == nil {
+				finalProposal = append(finalProposal, resultBytes)
+			} else {
+				panic(fmt.Errorf("invalid ban transaction in prepareProposal"))
```
Comment (sergio-mena) [inline abci/app.go]:
```suggestion
				panic(fmt.Errorf("ban transaction failed to marshal in prepareProposal"))
```

File: abci/app.go [hunk 26]
```diff
@@ -0,0 +1,377 @@
...
+	for i, tx := range processproposal.Txs {
+		if isBanTx(tx) {
+			var parsedBan model.BanTx
+			err := json.Unmarshal(tx, &parsedBan)
+			if err != nil {
+				return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_REJECT}, nil
+			}
+			bannedUsers[parsedBan.UserName] = struct{}{}
+		} else {
+			finishedBanTxIdx = i
```
Comment (sergio-mena) [inline abci/app.go:208]: Are we sure it's `i` and not `i+1`?

Comment (jmalicevic) [reply abci/app.go:208]: Checked in Go playground :D

Comment (sergio-mena) [reply abci/app.go:208]: Forget this comment... it's late and I didn't have coffee today

File: abci/app.go [hunk 27]
```diff
@@ -0,0 +1,377 @@
...
+			break
+		}
+	}
+
+	for _, tx := range processproposal.Txs[finishedBanTxIdx:] {
+		msg, err := model.ParseMessage(tx)
```
Comment (sergio-mena) [inline abci/app.go:217]: You first need to check `isBanTx` and it it is a ban Tx, you reject

Comment (jmalicevic) [reply abci/app.go:217]: If it is a `BanTx` it will fail to parse using `ParseMessage` because I don't append a string but rather create a completely different transaction type.

File: abci/app.go [hunk 28]
```diff
@@ -0,0 +1,377 @@
...
+
+		if isBanTx(tx) {
+			banTx := new(model.BanTx)
+			err = json.Unmarshal(tx, &banTx)
+			if err != nil {
+				respTxs[i] = &types.ExecTxResult{Code: CodeTypeEncodingError}
+			} else {
+				err := UpdateOrSetUser(app.state.DB, banTx.UserName, true, app.onGoingBlock)
+				if err != nil {
+					panic(err)
```
Comment (sergio-mena) [inline abci/app.go:244]: are we sure we want to panic here, rather than report an error in `ExecTxResult`?

File: abci/app.go [hunk 29]
```diff
@@ -0,0 +1,377 @@
...
+				panic(err)
+			}
+			// Append messages to chat history
+			app.onGoingBlock.Set([]byte("history"), []byte(chatHistory))
+			// This adds the user to the DB, but the data is not committed nor persisted until Comit is called
+			respTxs[i] = &types.ExecTxResult{Code: abci.CodeTypeOK}
+			app.state.Size++
+		}
+	}
+	app.state.Height = req.Height
```
Comment (sergio-mena) [inline abci/app.go:288]: We should probably check that the request height is the app's height + 1, as a validation at the beginning of this function

File: abci/app.go [hunk 30]
```diff
@@ -0,0 +1,377 @@
...
+}
+
+func (app *ForumApp) getWordsFromVe(voteExtensions []abci.ExtendedVoteInfo) string {
+	curseWordMap := make(map[string]int)
+	for _, vote := range voteExtensions {
+
+		// This code gets the curse words and makes sure that we do not add them more than once
+		// Thus ensuring each validator only adds one word once
+		voteExtensionString := string(vote.GetVoteExtension())
+		voteExtensionString = DedupWords(voteExtensionString)
```
Comment (sergio-mena) [inline abci/app.go]: You shouldn't need to dedup here... if there are duplicates here, you should panic?

Comment (sergio-mena) [reply abci/app.go]: I mean, this is called from `PrepareProposal`, and the vote extensions are the ones we gathered locally. So they should all have passed `VerifyVoteExtension`

**Review by sergio-mena — APPROVED** [review]
Awesome work, @jmalicevic !

File: abci/app.go (hunk 30 above)
Comment (jmalicevic) [reply abci/app.go]: You are right, this would make sense in `ProcessProposal` to validate the proposer did not mess with the count, but for that to make sense we would need to add gossiping of the vote extensions..so I will remove this from here to simplofy,.

File: abci/app.go (hunk 27 above)
Comment (sergio-mena) [reply abci/app.go:217]: Please add a comment explaining it (it's not clear IMO)

Comment (jmalicevic) [reply abci/app.go:217]: Done.
